storage.go

```diff
@@ -109,26 +111,26 @@ func (s *Storage) createDir(ctx context.Context, path string, opt pairStorageCre
 	o.Mode = ModeDir
 	o.ID = rp
 	o.Path = path
-	o.SetEtag(aws.StringValue(output.ETag))
+	o.SetEtag(*aws.String(*output.ETag))
```

Comment:
First of all, why do we need `aws.StringValue` here?

`output.ETag` is an `*string`, and it could be `nil`. `*output.ETag` could result in a panic. To be safe, we need to write in this way:

```go
if output.ETag != nil {
    o.SetEtag(*output.ETag)
}
```

As you can see, it adds an extra burden for reading/understanding our code. So it's common for golang projects to add convert-like functions, in which we will do this check and return empty value if the input point is nil. For example:

```go
func StringValue(s *string) string {
    if s == nil {
        return ""
    }
    return *s
}
```

So we can write in this way:

```go
o.SetEtag(aws.StringValue(output.ETag))
```

For aws-sdk-go-v2, `aws.StringValue` has been migrated to `aws.ToString`.

Compared the following two version, which way you prefer?

Version 1:

```go
o.SetEtag(*aws.String(*output.ETag))
```

Version 2:

```go
o.SetEtag(aws.ToString(output.ETag))
```

Our project intends to be a `production ready` library, please already take our code seriously. Try our best to avoid panic in runtime.

---

Another question is about readability. It's our duty to keep our code maintainable.

Let's speaking of this case:

```go
string(*output.SSEKMSKeyId)
```

Do we really need this conversion? Can we do it better?

Reply:
For **o.SetEtag(*aws.String(*output.ETag))**, **o.SetEtag(aws.ToString(output.ETag))** is much better. I will check all problems like this type.
Actually we don't need a convert like string(*output.SSEKMSKeyId). It can be **if v := *output.SSEKMSKeyId**;

Reply:
>  Actually we don't need a convert like **string(*output.SSEKMSKeyId)**. It can be if v := *output.SSEKMSKeyId.

`*output.SSEKMSKeyId` is not a safe operation too, please considering using `aws.ToString`.

Reply:
Yes, i got it. All the usage of aws.StringValue should consider that.